Mixer validator

mariia-iureva opened the pull request:
This validator ensures that the Mixer job configuration is correct, data is properly aligned, and filters are valid before starting the actual Mixer job.

- Here's[ the ticket](https://github.com/orgs/allenai/projects/137/views/1?pane=issue&itemId=78659489)
- Here's [a sample output](https://docs.google.com/document/d/1JGoiEujXxs33za6M8e5C1AO2lkBsf-tB4NhAi24L3cU/edit?usp=sharing)

**Configuration Validation**

- Loads and validates the configuration file structure
- Checks for required fields and correct data types in the config

**S3 Path and Permission Validation**

- Validates S3 paths for documents, attributes, and output
- Checks permissions (read/write) for the specified S3 paths
- Verifies the existence of parent directories for output paths

**Stream Filter Validation**

- Validates filter expressions in the configuration
- Checks for syntax errors and provides warnings for potential issues

**Document and Attribute Alignment**

- Samples a specified number of documents and their corresponding attribute files
- Validates the alignment between document and attribute files:
          - Checks that document and attribute files have the same number of lines
          - Verifies that both document and attribute files are valid JSONL
          - Ensures required fields are present in both document and attribute files

**Filter Execution Simulation**

- Executes filter commands on a sample of attribute files
- Reports on the number of lines that pass or are excluded by the filters
- Identifies any errors encountered during filter execution

**Attribute Name Validation**

- Cross-references attribute names used in filters with those found in attribute files
- Identifies missing attributes and potential typos in attribute names
- Provides a comprehensive list of all attributes found in the files

**File Sampling and Analysis**

- Downloads and analyzes a sample of files from the specified S3 paths
- Counts lines in both document and attribute files
- Validates the structure and content of sampled files

**Reporting and Logging**

- Provides detailed console output for each validation step
- Warns about potential issues or misconfigurations
- Summarizes the results of each validation phase

**Error Handling and Cleanup**

- Handles errors gracefully and provides informative error messages
- Cleans up temporary files and directories after validation

Whattabatt commented:
You're going to need to add this file's dependencies to pyproject.toml in order for it to run in a clean environment

Whattabatt commented on scripts/validate_mixer.py:
```diff
@@ -0,0 +1,935 @@
+    if config is None:
+        print("Configuration loading or validation failed")
+
+    if not validate_s3_paths_and_permissions(config):
```
I would consider pulling s3 related functions to a separate file to help legibility

mariia-iureva replied:
As I was creating a folder anyway, I split the whole script into smaller files with functionally grouped functions

Whattabatt commented on scripts/validate_mixer.py:
```diff
@@ -0,0 +1,935 @@
+        if field not in stream:
+            errors.append(f"Stream {index}: Missing required field: {field}")
+
+    if 'documents' in stream and not isinstance(stream['documents'], list):
```
Worth making a helper function to reduce a bit of redundant code, like check_type_if_present(structure,type,errors)

Whattabatt replied:
Another option would be to create a pydantic class or other validation config for each structure being validated, but I think with the fairly limited number of times you're currently using this pattern, it might be a little overkill.

mariia-iureva replied:
I added "check_type_if_present" function, now it lives in config_handler.py line 71

mariia-iureva commented:
> You're going to need to add this file's dependencies to pyproject.toml in order for it to run in a clean environment

Addressed this one and added dependencies

Whattabatt commented on scripts/validate_mixer.py:
Remove this?

Whattabatt commented:
The warnings produce a lot of noise, it'd be good to accept a 'verbose' flag and only log the warnings if it's set to true.

mariia-iureva commented:
> The warnings produce a lot of noise, it'd be good to accept a 'verbose' flag and only log the warnings if it's set to true.

Added --verbose flag and hid most of the print statements in it

mariia-iureva commented on scripts/validate_mixer/config_handler.py:
```diff
@@ -0,0 +1,105 @@
+        errors.extend(filter_errors)
+    return errors
+
+def check_type_if_present(stream: Dict[str, Any], field: str, expected_type: Union[Type, List[Type]], stream_index: int) -> List[str]:
```
Done

mariia-iureva commented on scripts/validate_mixer/validator.py:
```diff
@@ -0,0 +1,192 @@
+    
+    return all_valid
+
+# def validate_documents_and_attributes(config: Dict[str, Any], num_samples: int) -> bool:
```
Done

mariia-iureva commented on scripts/validate_mixer.py:
```diff
@@ -0,0 +1,935 @@
+    bucket, key = s3_path.replace("s3://", "").split("/", 1)
+    s3_client.download_file(bucket, key, local_path)
+
+# For local testing
```
Just deleted it

Whattabatt reviewed (approved):
You should delete the old validate_mixer.py script but otherwise looks good!
